src/backend/distributed/planner/recursive_planning.c

```diff
@@ -252,10 +309,516 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
+ */
+static void
+RecursivelyPlanNonColocatedSubqueries(Query *subquery, RecursivePlanningContext *context)
+{
+	NonColocatedSubqueryContext nonColocatedPlanningContext;
+
+	Query *anchorSubquery = NULL;
+	PlannerRestrictionContext *anchorPlannerRestrictionContext = NULL;
+	RelationRestrictionContext *anchorRelationRestrictionContext = NULL;
+	List *anchorRestrictionEquivalences = NIL;
+
+	FromExpr *joinTree = subquery->jointree;
+	PlannerRestrictionContext *plannerRestrictionContext = NULL;
+
+	/* we couldn't pick an anchor subquery, so we cannot continue */
+	anchorSubquery = AnchorSubquery(subquery);
+	if (anchorSubquery == NULL)
+	{
+		return;
+	}
+
+	plannerRestrictionContext = context->plannerRestrictionContext;
+
+	/* calculate the necessary information related to the anchor subquery */
+	anchorPlannerRestrictionContext =
+		FilterPlannerRestrictionForQuery(plannerRestrictionContext, anchorSubquery);
+	anchorRelationRestrictionContext =
+		anchorPlannerRestrictionContext->relationRestrictionContext;
+	anchorRestrictionEquivalences =
+		GenerateAllAttributeEquivalences(anchorPlannerRestrictionContext);
+
+	/* fill the non colocated planning context */
+	nonColocatedPlanningContext.subquery = subquery;
+	nonColocatedPlanningContext.subqueryPlannerRestriction = plannerRestrictionContext;
+
+	nonColocatedPlanningContext.anchorRelationRestrictionList =
+		anchorRelationRestrictionContext->relationRestrictionList;
+	nonColocatedPlanningContext.anchorAttributeEquivalances =
+		anchorRestrictionEquivalences;
+
+	/* handle from clause subqueries first */
+	RecursivelyPlanNonColocatedJoinWalker((Node *) joinTree, &nonColocatedPlanningContext,
+										  context);
+
+	/* handle subqueries in WHERE clause */
+	RecursivelyPlanNonColocatedSubqueriesInWhere(subquery, &nonColocatedPlanningContext,
+												 context);
+}
+
+
+/*
+ * AnchorSubqueryRte gets a query and searches for a subquery within the join tree of
+ * the query such that we can use it as our anchor subquery during our planning.
+ *
+ * The function returns NULL if it cannot find a proper range table entry for our
+ * purposes.
+ */
+static Query *
+AnchorSubquery(Query *subquery)
+{
+	FromExpr *joinTree = subquery->jointree;
+	Relids joinRelIds = get_relids_in_jointree((Node *) joinTree, false);
+	int currentRTEIndex = -1;
+	RangeTblEntry *rteRelation = NULL;
+
+	/*
+	 * Pick a random anchor subquery (i.e., the first) for now. We might consider
+	 * picking a better rte as the anchor. For example, we could iterate on the
+	 * joinRelIds, and check which rteIndex has more distribution key equiality
+	 * with rteIndexes. For the time being, the current primitive approach helps
+	 * us in many cases.
+	 */
+	while ((currentRTEIndex = bms_next_member(joinRelIds, currentRTEIndex)) >= 0)
+	{
+		RangeTblEntry *rte = rt_fetch(currentRTEIndex, subquery->rtable);
+
+		/*
+		 * We always prefer subqueries if we can find a suitable one for our
+		 * purposes.
```

Comment:
why do we prefer subqueries over relations? If you have `table JOIN subquery ON (non partition column)` wouldn't that be enabled by picking the table as an anchor?